src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -321,6 +327,343 @@ public async Task TestRemoteWorkspaceCircularReferences()
+
+                var localProject = localWorkspace.CurrentSolution.Projects.Single();
+                var remoteProject = remoteWorkspace.CurrentSolution.Projects.Single();
+
+                // Run generators locally
+                throwIfCalled = false;
+                var localCompilation = await localProject.GetCompilationAsync();
+
+                // Now run them remotely.  This must not actually call into the generator since nothing has changed.
+                throwIfCalled = true;
+                var remoteCompilation = await remoteProject.GetCompilationAsync();
+
+                await AssertSourceGeneratedDocumentsAreSame(localProject, remoteProject, expectedCount: sourceTexts.Length);
+            }
+        }
+
+        private static async Task AssertSourceGeneratedDocumentsAreSame(Project localProject, Project remoteProject, int expectedCount)
+        {
+            var localGeneratedDocs = (await localProject.GetSourceGeneratedDocumentsAsync()).ToImmutableArray();
+            var remoteGeneratedDocs = (await remoteProject.GetSourceGeneratedDocumentsAsync()).ToImmutableArray();
+
+            Assert.Equal(localGeneratedDocs.Length, remoteGeneratedDocs.Length);
+            Assert.Equal(expectedCount, localGeneratedDocs.Length);
+
+            for (var i = 0; i < expectedCount; i++)
+            {
+                var localDoc = localGeneratedDocs[i];
+                var remoteDoc = remoteGeneratedDocs[i];
+
+                Assert.Equal(localDoc.HintName, remoteDoc.HintName);
+                Assert.Equal(localDoc.DocumentState.Id, remoteDoc.DocumentState.Id);
+
+                var localText = await localDoc.GetTextAsync();
+                var remoteText = await localDoc.GetTextAsync();
+                Assert.Equal(localText.ToString(), remoteText.ToString());
+                Assert.Equal(localText.Encoding, remoteText.Encoding);
+                Assert.Equal(localText.ChecksumAlgorithm, remoteText.ChecksumAlgorithm);
+            }
+        }
+
+        private static SourceText CreateText(string content, Encoding encoding = null, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1)
+            => SourceText.From(content, encoding ?? Encoding.UTF8, checksumAlgorithm);
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree1(bool syncWithRemoteServer)
+        {
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", CreateText(Guid.NewGuid().ToString()))));
+        }
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree2(bool syncWithRemoteServer)
+        {
+            var sourceText = CreateText(Guid.NewGuid().ToString());
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", sourceText)),
+                ImmutableArray.Create(("SG.cs", sourceText)));
```

Comment:
this is actually simulating the SG producing the same results two times in a row.